File: 2024/python/day10/solution.py
def compile(instructions):
    pointer = 0
    registry = {}

    while pointer < len(instructions):
        instruction = instructions[pointer]
        params = instruction.split(" ")
        action = params[0]
        x = params[1]
        y = None if len(params) <= 2 else params[2]
        if action == "MOV":
            registry[y] = registry[x] if x in registry else int(x)
            pointer += 1
        if action == "INC":
            registry[x] = registry.get(x, 0) + 1
            pointer += 1
        if action == "DEC":
            registry[x] = registry.get(x, 0) - 1
            pointer += 1
        if action == "JMP":
            if x not in registry:
                registry[x] = 0
            value = registry[x]
            if value == 0:
                pointer = int(y)
            else:
                pointer += 1
    return  registry.get("A")

File: 2024/js/day10/solution.js
/**
 * @param {string[]} instructions - The instructions to execute
 * @returns {number} The value of the register A
 */
function compile(instructions) {
  var pointer = 0;
  var registry = {};

  while (pointer < instructions.length) {
    const instruction = instructions[pointer];
    const [action, x, y] = instruction.split(" ");

    if (action === "MOV") {
      if (x in registry) {
        registry[y] = registry[x];
      } else {
        registry[y] = Number(x);
      }
      pointer++;
    }

    if (action === "INC") {
      registry[x] = registry[x] || 0;
      registry[x]++;
      pointer++;
    }

    if (action === "DEC") {
      registry[x] = registry[x] || 0;
      registry[x]--;
      pointer++;
    }

    if (action == "JMP") {
      let value = registry[x] || 0;
      if (value === 0) {
        pointer = Number(y);
      } else {
        pointer++;
      }
    }
  }
  return registry["A"];
}
